okhttp/src/main/java/okhttp3/internal/http/BridgeInterceptor.kt

```diff
@@ -0,0 +1,120 @@
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package okhttp3.internal.http
+
+import java.io.IOException
+import okhttp3.Cookie
+import okhttp3.CookieJar
+import okhttp3.Interceptor
+import okhttp3.Response
+import okhttp3.internal.Version
+import okio.GzipSource
+
+import okhttp3.internal.Util.hostHeader
+import okio.buffer
+
+/**
+ * Bridges from application code to network code. First it builds a network request from a user
+ * request. Then it proceeds to call the network. Finally it builds a user response from the network
+ * response.
+ */
+class BridgeInterceptor(private val cookieJar: CookieJar) : Interceptor {
+
+  @Throws(IOException::class)
+  override fun intercept(chain: Interceptor.Chain): Response {
+    val userRequest = chain.request()
+    val requestBuilder = userRequest.newBuilder()
+
+    val body = userRequest.body()
+    body?.let {
+      val contentType = body.contentType()
+      contentType?.let {
+        requestBuilder.header("Content-Type", contentType.toString())
+      }
+
+      val contentLength = body.contentLength()
+      if (contentLength != -1L) {
+        requestBuilder.header("Content-Length", contentLength.toString())
+        requestBuilder.removeHeader("Transfer-Encoding")
+      } else {
+        requestBuilder.header("Transfer-Encoding", "chunked")
+        requestBuilder.removeHeader("Content-Length")
+      }
+    }
+
+    if (userRequest.header("Host") == null) {
+      requestBuilder.header("Host", hostHeader(userRequest.url(), false))
+    }
+
+    if (userRequest.header("Connection") == null) {
+      requestBuilder.header("Connection", "Keep-Alive")
+    }
+
+    // If we add an "Accept-Encoding: gzip" header field we're responsible for also decompressing
+    // the transfer stream.
+    var transparentGzip = false
+    if (userRequest.header("Accept-Encoding") == null && userRequest.header("Range") == null) {
+      transparentGzip = true
+      requestBuilder.header("Accept-Encoding", "gzip")
+    }
+
+    val cookies = cookieJar.loadForRequest(userRequest.url())
+    if (!cookies.isEmpty()) {
+      requestBuilder.header("Cookie", cookieHeader(cookies))
+    }
+
+    if (userRequest.header("User-Agent") == null) {
+      requestBuilder.header("User-Agent", Version.userAgent())
+    }
+
+    val networkResponse = chain.proceed(requestBuilder.build())
+
+    HttpHeaders.receiveHeaders(cookieJar, userRequest.url(), networkResponse.headers())
+
+    val responseBuilder = networkResponse.newBuilder()
+        .request(userRequest)
+
+    if (transparentGzip
+        && "gzip".equals(networkResponse.header("Content-Encoding"), ignoreCase = true)
+        && HttpHeaders.hasBody(networkResponse)) {
+
+      networkResponse.body()?.let {
+        val responseBody = GzipSource(it.source())
```

Comment:
nit: introduce the `body` variable and change to `body.source()`?

Reply:
The `body` variable name is shadowed by previous code,so I use the name `responseBody`  and changed `val responseBody = GzipSource(responseBody.source())` to  `val gzipSource = GzipSource(responseBody.source())`,it looks more clear.